### ISSUE TITLE
Timeouts trigger one tick late when there's a delay between creating and running the event loop

### ISSUE THREAD

**Issue by xalri** [carllerche/mio#383]
This first manifested through mioco as https://github.com/dpc/mioco/issues/126, but it seems to be a bug in mio itself.

Here's what I _think's_ happening:
At the end of a call to `Timer::tick_to(t)` (https://github.com/carllerche/mio/blob/master/src/timer.rs#L202) where there are no timouts set to trigger on tick `t`, `self.tick` is equal to `t + 1`, one tick past the current time.
```rust
    while self.tick <= now {
        ...
        if curr == EMPTY {
            self.tick += 1;
```

Then, `next_tick_in_ms` returns the time in ms to the start of `self.tick + 1` (`== t + 2`), so any timeouts on tick `t + 1` don't trigger untill tick `t + 2`. I'm not sure why, but the effects of this are only visible when there's a delay between creating and starting the event loop. I made a small change to `tick_to` to test this, and it fixed the example below (https://gist.github.com/xalri/bc54030fa7ca48496fa253586d6f6a99).

Anyway, The effect is that timeouts trigger one tick late.
Here's an example which should timeout every 100ms (the default timer resolution), but, with a 10ms delay before starting the loop, the first takes 290ms, and all subsequent timeouts take 200ms.

``` rust
#[macro_use] extern crate log;
extern crate mio;

mod logger;
use std::time::Duration;

struct Timer;
impl mio::Handler for Timer{
    type Timeout = ();
    type Message = ();

    fn timeout(&mut self, evl: &mut mio::EventLoop<Timer>, _: ()){
		info!("timeout");
        evl.timeout((),Duration::from_millis(100)).unwrap();
    }
}

fn main() {
    logger::init().unwrap();
    let mut event_loop = mio::EventLoop::new().unwrap();
    std::thread::sleep(Duration::from_millis(10));
    event_loop.timeout((),Duration::from_millis(100)).unwrap();
    event_loop.run(&mut Timer{}).unwrap();
}
```
And the output, using mio master with a little more debug info:
```
mio::poll       32263544ms  registering with poller
mio::event_loop 32263554ms  event_loop.timeout Duration { secs: 0, nanos: 100000000 }
mio::timer      32263554ms  timeout_ms now: 32263553   delay: 100
mio::timer      32263554ms  timeout_at_ms at: 32263653
mio::timer      32263554ms  timeout_at_ms at-self.start: 110
mio::timer      32263554ms  timeout_at_ms tick: 2, self.tick: 0
mio::timer      32263554ms  inserted timout; slot=2; token=Token(0)
mio::event_loop 32263554ms  event loop tick
mio::timer      32263554ms  next_tick_in_ms  self.tick: 0, self.tick_ms: 100
mio::timer      32263554ms  next_tick_in_ms  now: 32263554, next: 32263643, nxt-now: 89
mio::event_loop 32263643ms  io_process(..); cnt=0; len=0
mio::timer      32263643ms  tick_to; now=1; tick=0
mio::timer      32263643ms  ticking; curr=Token(18446744073709551615)
mio::timer      32263643ms  ticking; curr=Token(18446744073709551615)
mio::event_loop 32263643ms  event loop tick
mio::timer      32263643ms  next_tick_in_ms  self.tick: 2, self.tick_ms: 100
mio::timer      32263643ms  next_tick_in_ms  now: 32263643, next: 32263843, nxt-now: 200
mio::event_loop 32263844ms  io_process(..); cnt=0; len=0
mio::timer      32263844ms  tick_to; now=3; tick=2
mio::timer      32263844ms  ticking; curr=Token(0)
mio::timer      32263844ms  triggering; token=Token(0)
mio::timer      32263844ms  unlinking timeout; slot=2; token=Token(0)
mio             32263844ms  timeout
mio::event_loop 32263844ms  event_loop.timeout Duration { secs: 0, nanos: 100000000 }
mio::timer      32263844ms  timeout_ms now: 32263843   delay: 100
mio::timer      32263844ms  timeout_at_ms at: 32263943
mio::timer      32263844ms  timeout_at_ms at-self.start: 400
mio::timer      32263844ms  timeout_at_ms tick: 4, self.tick: 2
mio::timer      32263844ms  inserted timout; slot=4; token=Token(0)
mio::timer      32263844ms  tick_to; now=3; tick=2
mio::timer      32263844ms  ticking; curr=Token(18446744073709551615)
mio::timer      32263844ms  ticking; curr=Token(18446744073709551615)
mio::event_loop 32263844ms  event loop tick
mio::timer      32263844ms  next_tick_in_ms  self.tick: 4, self.tick_ms: 100
mio::timer      32263844ms  next_tick_in_ms  now: 32263843, next: 32264043, nxt-now: 200
mio::event_loop 32264044ms  io_process(..); cnt=0; len=0
mio::timer      32264044ms  tick_to; now=5; tick=4
mio::timer      32264044ms  ticking; curr=Token(0)
mio::timer      32264044ms  triggering; token=Token(0)
mio::timer      32264044ms  unlinking timeout; slot=4; token=Token(0)
mio             32264044ms  timeout
```

**Comment by rrichardson** [comment]
I refactored Timer quite a bit in https://github.com/carllerche/mio/pull/382/commits/e8280e9e3c1bee8f72b269296a8ef13dc098e8a4 

I'll run your test against it to see if it improves things.

**Comment by tmiasko** [comment]
Currently implementation from master branch seems to trigger the timeout either
after 100 ms or 200 ms.

The behaviour is a little more consistent if obtaining current tick is changed
to round to the nearest tick in the same way as it is done when scheduling a
new timeout. This usually triggers the timeout after 100 ms, and in very rare
cases almost immediately. What would you say about this solution?

**Comment by carllerche** [comment]
Thanks for the report. I ran your repro and it seems like with the timer changes that have been already merged the current behavior is correct.

If there is still a bug, please open a new ticket :)
